In [1]:
%%HTML
<h2 style = "color:blue"> Projeto Mega - V1.JF </h2>
<img src = "imagens/LogoJoao-2019.png" alt="Nada" width=80 height=80>
    <p> Projeto inicial de estudo de Jogos da Mega Senna.</p>

<h4> Primeiros passos Criação do Dataframe Pandas</h4>
<p> Importação do arquivo xlx para dataframe pandas e primeiros filtros aplicados </p>


In [2]:
# imports iniciais 
import numpy as np
import pandas as pd
import seaborn as sns
from matplotlib import pyplot as plt
from datetime import datetime

%matplotlib inline

In [3]:
#Dataset
data_path = "PlanilhaBruta01.xlsx"
df = pd.read_excel(data_path)

In [4]:
df.head(7)

Concurso  Data Sorteio  1ª Dezena  2ª Dezena  3ª Dezena  4ª Dezena  \
0       1.0       35135.0       41.0        5.0        4.0       52.0   
1       2.0       35142.0        9.0       39.0       37.0       49.0   
2       3.0       35149.0       36.0       30.0       10.0       11.0   
3       NaN           NaN        NaN        NaN        NaN        NaN   
4       4.0       35156.0        6.0       59.0       42.0       27.0   
5       5.0       35163.0        1.0       19.0       46.0        6.0   
6       6.0       35170.0       19.0       40.0        7.0       13.0   

   5ª Dezena  6ª Dezena  Arrecadacao_Total  Ganhadores_Sena  \
0       30.0       33.0                0.0              0.0   
1       43.0       41.0                0.0              1.0   
2       29.0       47.0                0.0              2.0   
3        NaN        NaN                NaN              NaN   
4        1.0        5.0                0.0              0.0   
5       16.0        2.0                0.0              0.0   
6       22.0       47.0                0.0              0.0   

             ...              UF Rateio_Sena  Ganhadores_Quina  Rateio_Quina  \
0            ...             NaN        0.00              17.0      39158.92   
1            ...              PR  2307162.23              65.0      14424.02   
2            ...              RN   391192.51              62.0      10515.93   
3            ...              SP         NaN               NaN           NaN   
4            ...             NaN        0.00              39.0      15322.24   
5            ...             NaN        0.00              98.0       5318.10   
6            ...             NaN        0.00             109.0       7214.66   

   Ganhadores_Quadra  Rateio_Quadra  Acumulado Valor_Acumulado  \
0             2016.0         330.21        SIM      1714650.23   
1             4488.0         208.91        NÃO            0.00   
2             4261.0         153.01        NÃO            0.00   
3                NaN            NaN        NaN             NaN   
4             3311.0         180.48        SIM       717080.75   
5             5399.0          96.53        SIM      1342488.85   
6             7147.0         110.03        SIM      2286166.33   

   Estimativa_Prêmio  Acumulado_Mega_da_Virada  
0                0.0                       0.0  
1                0.0                       0.0  
2                0.0                       0.0  
3                NaN                       NaN  
4                0.0                       0.0  
5                0.0                       0.0  
6                0.0                       0.0  

[7 rows x 21 columns]

In [5]:
#Informacoes iniciais do dataframe
list(df)

['Concurso',
 'Data Sorteio',
 '1ª Dezena',
 '2ª Dezena',
 '3ª Dezena',
 '4ª Dezena',
 '5ª Dezena',
 '6ª Dezena',
 'Arrecadacao_Total',
 'Ganhadores_Sena',
 'Cidade',
 'UF',
 'Rateio_Sena',
 'Ganhadores_Quina',
 'Rateio_Quina',
 'Ganhadores_Quadra',
 'Rateio_Quadra',
 'Acumulado',
 'Valor_Acumulado',
 'Estimativa_Prêmio',
 'Acumulado_Mega_da_Virada']

In [6]:
#Mais informações 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2439 entries, 0 to 2438
Data columns (total 21 columns):
Concurso                    2171 non-null float64
Data Sorteio                2171 non-null float64
1ª Dezena                   2171 non-null float64
2ª Dezena                   2171 non-null float64
3ª Dezena                   2171 non-null float64
4ª Dezena                   2171 non-null float64
5ª Dezena                   2171 non-null float64
6ª Dezena                   2171 non-null float64
Arrecadacao_Total           2171 non-null float64
Ganhadores_Sena             2171 non-null float64
Cidade                      464 non-null object
UF                          750 non-null object
Rateio_Sena                 2171 non-null float64
Ganhadores_Quina            2171 non-null float64
Rateio_Quina                2171 non-null float64
Ganhadores_Quadra           2171 non-null float64
Rateio_Quadra               2171 non-null float64
Acumulado                   2171 non-null objec

In [7]:
# Filtro - separação dos dados para estudo 
new_df = df.iloc[:, [0,2,3,4,5,6,7,9,12]] #Comando para separar as colunas que necessito inicialmente no estudo

In [8]:
new_df.head(10)

Concurso  1ª Dezena  2ª Dezena  3ª Dezena  4ª Dezena  5ª Dezena  6ª Dezena  \
0       1.0       41.0        5.0        4.0       52.0       30.0       33.0   
1       2.0        9.0       39.0       37.0       49.0       43.0       41.0   
2       3.0       36.0       30.0       10.0       11.0       29.0       47.0   
3       NaN        NaN        NaN        NaN        NaN        NaN        NaN   
4       4.0        6.0       59.0       42.0       27.0        1.0        5.0   
5       5.0        1.0       19.0       46.0        6.0       16.0        2.0   
6       6.0       19.0       40.0        7.0       13.0       22.0       47.0   
7       7.0       56.0       38.0       21.0       20.0        3.0        5.0   
8       8.0       53.0       17.0       38.0        4.0       47.0       37.0   
9       9.0       55.0       43.0       56.0       54.0        8.0       60.0   

   Ganhadores_Sena  Rateio_Sena  
0              0.0         0.00  
1              1.0   2307162.23  
2              2.0    391192.51  
3              NaN          NaN  
4              0.0         0.00  
5              0.0         0.00  
6              0.0         0.00  
7              0.0         0.00  
8              0.0         0.00  
9              0.0         0.00

In [9]:
new_df.tail(5)

Concurso  1ª Dezena  2ª Dezena  3ª Dezena  4ª Dezena  5ª Dezena  \
2434    2167.0       37.0       45.0       38.0       43.0       54.0   
2435    2168.0       37.0       25.0       29.0        4.0        1.0   
2436    2169.0       45.0        7.0       34.0       59.0       51.0   
2437    2170.0       36.0       21.0       51.0       10.0       24.0   
2438    2171.0       19.0       12.0       13.0       36.0       55.0   

      6ª Dezena  Ganhadores_Sena  Rateio_Sena  
2434       27.0              0.0         0.00  
2435       27.0              1.0   8177021.67  
2436       54.0              0.0         0.00  
2437       38.0              0.0         0.00  
2438       44.0              1.0  21978571.91

In [10]:
# Planilha contendo apenas colunas que interessam ao primeiro estudo salvo no arquivoCSV abaixo
new_df.to_csv("data/MegaPrimeiroFiltro.csv")

In [11]:
# Limpeza inicial de dados NaN
# Identificando NaN
new_df.isnull().sum()

Concurso           268
1ª Dezena          268
2ª Dezena          268
3ª Dezena          268
4ª Dezena          268
5ª Dezena          268
6ª Dezena          268
Ganhadores_Sena    268
Rateio_Sena        268
dtype: int64

<h2 style="color:blue"> Conclusão do filtro NaN acima:</h2>
<p> Em cada coluna do dataframe existem 268 linhas contendo dados missing NaN.<br>
 Desta forma serão retirados do dataframe, pois representam jogos que, de alguma forma, os dados não foram informandos pela Caixa Econômica.<br>
 Como os NaN aparecem em todos as linhas das colunas selecionadas, a retirada é a melhor alternativa para limpar o dataframe da falta de informação.<br></p>
 


In [12]:
#Exclusão dos NaN 
new_df.dropna(inplace=True)

/home/joaoflorentino/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [13]:
new_df.head(10)

Concurso  1ª Dezena  2ª Dezena  3ª Dezena  4ª Dezena  5ª Dezena  \
0        1.0       41.0        5.0        4.0       52.0       30.0   
1        2.0        9.0       39.0       37.0       49.0       43.0   
2        3.0       36.0       30.0       10.0       11.0       29.0   
4        4.0        6.0       59.0       42.0       27.0        1.0   
5        5.0        1.0       19.0       46.0        6.0       16.0   
6        6.0       19.0       40.0        7.0       13.0       22.0   
7        7.0       56.0       38.0       21.0       20.0        3.0   
8        8.0       53.0       17.0       38.0        4.0       47.0   
9        9.0       55.0       43.0       56.0       54.0        8.0   
10      10.0       25.0        4.0       18.0       57.0       21.0   

    6ª Dezena  Ganhadores_Sena  Rateio_Sena  
0        33.0              0.0         0.00  
1        41.0              1.0   2307162.23  
2        47.0              2.0    391192.51  
4         5.0              0.0         0.00  
5         2.0              0.0         0.00  
6        47.0              0.0         0.00  
7         5.0              0.0         0.00  
8        37.0              0.0         0.00  
9        60.0              0.0         0.00  
10       38.0              0.0         0.00

In [14]:
#verificação Dropna funcionou
new_df.isnull().sum()

Concurso           0
1ª Dezena          0
2ª Dezena          0
3ª Dezena          0
4ª Dezena          0
5ª Dezena          0
6ª Dezena          0
Ganhadores_Sena    0
Rateio_Sena        0
dtype: int64

<h3 style="Color:red"> Remoção dos dados NaN </h3>
<p> Podemos ver que os NaN foram removidos do dataframe e agora podemos continuar com valores para análise preliminar. <br>
Em seguida vamos verificar quantos, dos jagos atuais, apresentaram ganhadores.</p>

In [15]:
# filtrando os sorteios nos quais houveram ganhadores 
# Usando drop() onde a coluna Ganhadores_Sena é igual a zero
indice = new_df.loc[(new_df['Ganhadores_Sena'] == 0)] #Aqui separo os que não ganharam 
df2 = new_df.drop(indice.index) # aqui crio novo dataframe somente com concursos ganhadores.


In [16]:
# Mostrando dataframe  com os ganhadores 
df2.head(10)

Concurso  1ª Dezena  2ª Dezena  3ª Dezena  4ª Dezena  5ª Dezena  \
1        2.0        9.0       39.0       37.0       49.0       43.0   
2        3.0       36.0       30.0       10.0       11.0       29.0   
11      11.0       25.0       15.0       58.0       37.0       59.0   
17      17.0       10.0       20.0        6.0       19.0       51.0   
24      24.0        1.0        8.0       14.0       28.0       33.0   
29      29.0       14.0       56.0       58.0        8.0       43.0   
33      33.0        5.0       33.0       49.0       17.0       42.0   
41      41.0       38.0       26.0       35.0       39.0       12.0   
49      49.0       33.0        7.0       43.0       35.0       13.0   
54      54.0        5.0       21.0        9.0       56.0       18.0   

    6ª Dezena  Ganhadores_Sena  Rateio_Sena  
1        41.0              1.0   2307162.23  
2        47.0              2.0    391192.51  
11       38.0              1.0  15591365.07  
17       13.0              1.0   6789869.08  
24       43.0              1.0  18661679.61  
29        3.0              1.0   5401793.60  
33       39.0              1.0  13997410.74  
41       47.0              1.0  17019238.50  
49       55.0              1.0   8405156.76  
54       36.0              2.0   7469653.99

In [17]:
# Salvando novo CSV contendo somente ganhadores sem NaN
df2.to_csv('data/somenteGanadores.csv')

### Avaliação dos dados do dataframe de ganhadores:
<p> A partir desta limpeza inicial do dataframe, quero fazer algumas descobertas: </p>
<ol>
    <li> Numero de concursos vencedores até a data da extração do dataframe </li>
    <li> Percentual de concursos vencedores </li>
    <li> Dezenas mais sorteadas </li>
    <li> Dezenas menos sorteadas </li>
    <li> Dezenas não sorteadas </li>
</ol>
<p> A partir destes levantamentos iniciais, vejamos o que já se pode concluir deste jogo de azar</p>

In [37]:
from IPython.core.display import HTML
display(HTML ('<h3 style= "color:blue"> Análise:</h3>'))
print ("1. Numero de ganhadores = ", len(df2))
print ()
print ("2. Percentual de jogos ganhos")
print ()
a = len(df)
tot = a - (df['Concurso'].isnull().sum())
b = len(df2)
per = b/tot
print("Numero Total de sorteios = ", a)
print("Numero Total de sorteios menos os NaN = ", tot)
print("Numero de Ganhadores = ", b)
print("Percentual de Ganhadores:{0:2.4f}%".format(per * 100)) # explicando a formatação dos numeros
#{0:2.2f}%  vai imprimir o primeiro numero do format (0) com duas casas decimiais no inicio e quatro no float e 
# depois o sinal de percentual. 

1. Numero de ganhadores =  505

2. Percentual de jogos ganhos

Numero Total de sorteios =  2439
Numero Total de sorteios menos os NaN =  2171
Numero de Ganhadores =  505
Percentual de Ganhadores:23.2612%


<p style ="color:red"> Considerações</p>
No calculo acima, foram retirados da equação os valores missing NaN. Desta forma tanto jogos ganhos quanto
acumulados, podem ter sido desconsiderados nos calculos devido a esta inconsistencia da informação.